projects/packages/post-list/src/class-admin.php

```diff
@@ -0,0 +1,228 @@
+			return $columns;
+		}
+
+		return array_merge( $columns, array( 'post-list-column' => 'WPAP' ) );
+	}
+
+	/**
+	 * Add the Jetpack Post List footer placeholder.
+	 */
+	public function add_admin_footer_placeholder() {
+		echo '<div class="post-list__notice-placeholder"></div>';
+	}
+
+	/**
+	 * Pick post data, parse and render in JSON format.
+	 *
+	 * @param string $column  The name of the column to display.
+	 * @param int    $post_id The current post ID.
+	 */
+	public function parse_and_render_post_data( $column, $post_id ) {
+		if ( ! self::is_posts_list() ) {
+			return;
+		}
+
+		$post = get_post( $post_id );
+
+		$featured_image = $this->get_featured_or_first_post_image( $post );
+
+		if ( 'post-list-column' === $column ) {
+			echo '<script type="application/json">';
+			echo wp_json_encode(
+				array(
+					'id'             => $post->ID,
+					'type'           => $post->post_type,
+					'status'         => $post->post_status,
+					'date_gmt'       => $post->post_date_gmt,
+					'statuses'       => self::get_post_statuses( $post_id ),
+					'featured_image' => $featured_image,
+				)
+			);
+			echo '</script>';
+		}
+	}
+
+	/**
+	 * Return post statuses definition object,
+	 * with especial cases such as `scheduled`, `pending`, etc.
+	 *
+	 * @param int $post_id The current post ID.
+	 * @return array Post statuses object.
+	 */
+	public function get_post_statuses( $post_id ) {
+		return array_merge(
+			array(
+				'publish' => __( 'Published', 'jetpack' ),
+				'private' => __( 'Published', 'jetpack' ),
+				'future'  => __( 'Scheduled', 'jetpack' ),
+				'pending' => __( 'Pending Review', 'jetpack' ),
+				'draft'   => __( 'Draft', 'jetpack' ),
+			),
+			get_post_statuses( $post_id )
+		);
+	}
+
+	/**
+	 * Enqueue scripts depending on the post-list query var.
+	 *
+	 * @param string $hook Page hook.
+	 */
+	public function enqueue_scripts( $hook ) {
+		if ( self::is_posts_list() && 'edit.php' === $hook ) {
+			$build_assets = require_once __DIR__ . '/../build/index.asset.php';
+			$plugin_path  = Assets::get_file_url_for_environment( '../build/index.js', '../build/index.js', __FILE__ );
+
+			wp_enqueue_script(
+				'jetpack_posts_list_ui_script',
+				$plugin_path,
+				$build_assets['dependencies'],
+				$build_assets['version'],
+				true
+			);
+
+			wp_enqueue_style(
+				'jetpack_posts_list_ui_style',
+				plugin_dir_url( __DIR__ ) . 'build/style-index.css',
+				array( 'wp-components' ),
+				$build_assets['version']
+			);
+
+			wp_style_add_data(
+				'jetpack_posts_list_ui_style',
+				'rtl',
+				plugin_dir_url( __DIR__ ) . 'build/index.rtl.css'
+			);
+		}
+	}
+
+	/**
+	 * Return the featured image or if no featured image is set, return the first image in the post. If neither exists
+	 * return the featured image array with null values.
+	 *
+	 * @param object $post The current post.
+	 * @return array The featured image id and URLs
+	 */
+	protected function get_featured_or_first_post_image( $post ) {
```

Comment:
Jetpack includes a class that could be really useful here: `Jetpack_PostImages`. Obviously it's not ideal that it's in the plugin instead of in a package so your package cannot just depend on it just yet, but that may be the push we need to pick this PR back up and move the class into its own package: #19579

Reply:
Yeah, that'd be acceptable I think, and copying and pasting should make it relatively easy to replace later on.

Reply:
Jetpack_PostImages looks good, but I'd like to propose we continue with the current code approach this iteration.

Post_Images::get_image() doesn't return the image/attachment ID or thumbnail. To get the thumbnail we call fit_image_url(src) and it uses Photon, which is cool, but with our current implementation, we can get an image from all the same sources (gallery, image, slideshow, featured image), their ID, and their 150x150 thumbnail from media library with minimal code and without the need of Photon or any other dependencies.

The main downside to our current approach is it relies on the img class 'wp-image-ID' existing, but I feel like that may actually be a pretty reliable marker. Further, we don't use this class marker to get the featured image thumbnail. So our primary image source doesn't have this concern.

Ultimately, in our case of finding a post thumbnail image (as a fallback to the featured image), I'm a fan of traversing the post's HTML directly and plucking the first image, rather than searching and compiling each possible source of an image. I feel like the DomXPath approach will be easier to unit test as it's just string manipulation. Using 'wp-image-ID' in our case is just opportunistic to make grabbing the thumbnail easy. But if we integrate Photon either way, we could just grab the first post img using DomXPath and Photon that.